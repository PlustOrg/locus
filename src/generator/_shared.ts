// Shared generator utilities (pure, no side effects). Extracted from scattered helpers.

export function pluralize(name: string): string {
  if (name.endsWith('y') && !/[aeiou]y$/i.test(name)) return name.slice(0, -1) + 'ies';
  if (name.endsWith('s')) return name + 'es';
  return name + 's';
}

export function kebabCasePageName(name: string): string {
  return name
    .replace(/([a-z])([A-Z])/g, '$1-$2')
    .replace(/\s+/g, '-')
    .replace(/_/g, '-')
    .toLowerCase();
}

export function sortByName<T extends { name: string }>(arr: T[]): T[] {
  return [...arr].sort((a,b)=> a.name.localeCompare(b.name));
}

export function stableJSONString(obj: any): string {
  // Deterministic stringify (simple: rely on JSON.stringify ordering of created literal with sorted keys)
  return JSON.stringify(sortKeysDeep(obj), null, 2);
}

function sortKeysDeep(value: any): any {
  if (Array.isArray(value)) return value.map(sortKeysDeep);
  if (value && typeof value === 'object') {
    const out: any = {};
    for (const k of Object.keys(value).sort()) out[k] = sortKeysDeep(value[k]);
    return out;
  }
  return value;
}
