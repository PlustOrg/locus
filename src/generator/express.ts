import { Entity } from '../ast';

function pluralize(name: string): string {
  if (name.endsWith('y') && !/[aeiou]y$/i.test(name)) return name.slice(0, -1) + 'ies';
  if (name.endsWith('s')) return name + 'es';
  return name + 's';
}
export function generateExpressApi(entities: Entity[], opts?: { pluralizeRoutes?: boolean }): Record<string, string> {
  const files: Record<string, string> = {};
  const mounts: string[] = [];
  const sorted = [...entities].sort((a, b) => a.name.localeCompare(b.name));
  for (const e of sorted) {
    const base = e.name.charAt(0).toLowerCase() + e.name.slice(1);
    const routeBase = opts?.pluralizeRoutes ? pluralize(base) : base;
  const route = `import { Router } from 'express'
import { PrismaClient } from '@prisma/client'
const prisma = new PrismaClient()
export const router = Router()

// GET /${routeBase}
router.get('/${routeBase}', async (req, res) => {
  try {
  const skip = req.query.skip !== undefined ? Number(req.query.skip) : undefined
  const take = req.query.take !== undefined ? Number(req.query.take) : undefined
    let where: any = undefined
    if (req.query.where) {
      try { where = typeof req.query.where === 'string' ? JSON.parse(req.query.where) : req.query.where } catch { return res.status(400).json({ error: 'Invalid where JSON' }) }
    }
  if (skip !== undefined && Number.isNaN(skip)) return res.status(400).json({ error: 'skip must be a number' })
  if (take !== undefined && Number.isNaN(take)) return res.status(400).json({ error: 'take must be a number' })
  const rows = await prisma.${base}.findMany({ where, skip, take })
    res.json(rows)
  } catch (err: any) {
    res.status(500).json({ error: err?.message || 'Server error' })
  }
})

// GET /${routeBase}/:id
router.get('/${routeBase}/:id', async (req, res) => {
  try {
    const id = Number(req.params.id)
  const row = await prisma.${base}.findUnique({ where: { id } })
    if (!row) return res.status(404).end()
    res.json(row)
  } catch (err: any) {
    res.status(500).json({ error: err?.message || 'Server error' })
  }
})

// POST /${routeBase}
router.post('/${routeBase}', async (req, res) => {
  try {
  // TODO(validation): derive schema from entity definition and validate body
  const created = await prisma.${base}.create({ data: req.body })
    res.status(201).json(created)
  } catch (err: any) {
    res.status(400).json({ error: err?.message || 'Create failed' })
  }
})

// PUT /${routeBase}/:id
router.put('/${routeBase}/:id', async (req, res) => {
  try {
    const id = Number(req.params.id)
  if (Number.isNaN(id)) return res.status(400).json({ error: 'id must be a number' })
  // TODO(validation): derive schema from entity definition and validate body
  const updated = await prisma.${base}.update({ where: { id }, data: req.body })
    res.json(updated)
  } catch (err: any) {
    res.status(400).json({ error: err?.message || 'Update failed' })
  }
})

// DELETE /${routeBase}/:id
router.delete('/${routeBase}/:id', async (req, res) => {
  try {
    const id = Number(req.params.id)
  if (Number.isNaN(id)) return res.status(400).json({ error: 'id must be a number' })
  await prisma.${base}.delete({ where: { id } })
    res.status(204).end()
  } catch (err: any) {
    res.status(400).json({ error: err?.message || 'Delete failed' })
  }
})
`;
  files[`routes/${base}.ts`] = route;
  mounts.push(base);
  }
  // Optional: basic app bootstrap
  // Use CommonJS style dynamic requires to avoid extension resolution issues under differing module loaders
  const imports = mounts.map(n => `const { router: ${n}Router } = require('./routes/${n}.ts')`).join('\n');
  const uses = mounts.map(n => `app.use('/${opts?.pluralizeRoutes ? pluralize(n) : n}', ${n}Router)`).join('\n');
  files['server.ts'] = `/* eslint-disable */
import express from 'express'
import cors from 'cors'
import path from 'path'
${imports}

const app = express()
app.use(express.json())
if (process.env.ENABLE_CORS === '1') { app.use(cors()) }
const publicDir = path.join(__dirname, 'next-app', 'public')
try { app.use(express.static(publicDir)) } catch {}
// Basic health/readiness endpoints
app.get('/healthz', (req, res) => { res.json({ ok: true, uptime: process.uptime(), ts: Date.now() }) })
let startedAt = Date.now();
app.get('/readyz', (req, res) => { res.json({ ready: true, uptime: process.uptime(), startedAt }) })
${uses}

export function startServer(port: number = Number(process.env.API_PORT || process.env.PORT) || 3001) {
  return app.listen(port, () => { console.log('[locus][api] listening on :' + port); console.log('[locus][api] ready'); })
}
export default app
// Auto-start when run directly (dev:api script)
if (require.main === module) {
  startServer();
}
`;
  return files;
}
