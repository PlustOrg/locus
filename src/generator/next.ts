export function generateNextApp(pages: Array<{ name: string }>): Record<string, string> {
  const files: Record<string, string> = {};
  // globals.css imports public/theme.css
  files['next-app/app/globals.css'] = `@import url('/theme.css');
body { margin: 0; font-family: system-ui, -apple-system, Segoe UI, Roboto, Ubuntu, Cantarell, Noto Sans, Helvetica Neue, Arial; }
`;  
  files['next-app/app/layout.tsx'] = `import './globals.css'
import React from 'react'
export const metadata = { title: 'Locus App' }
export default function RootLayout({ children }: { children: React.ReactNode }) {
  return (<html lang="en"><body data-theme="light">{children}</body></html>)
}
`;

  const sorted = [...pages].sort((a, b) => a.name.localeCompare(b.name));
  const kebab = (name: string) => name
    .replace(/([a-z])([A-Z])/g, '$1-$2')
    .replace(/\s+/g, '-')
    .replace(/_/g, '-')
    .toLowerCase();

  // Landing page listing links to each generated page
  if (sorted.length) {
    const links = sorted.map(p => `      <li><a href="/${kebab(p.name)}">${p.name}</a></li>`).join('\n');
    files['next-app/app/page.tsx'] = `import React from 'react'
export default function Landing(){
  return <main style={{padding:20,fontFamily:'system-ui'}}>
    <h1>Locus Pages</h1>
    <ul style={{lineHeight:1.8}}>
${links}
    </ul>
  </main>
}
`;
  } else {
    files['next-app/app/page.tsx'] = `export default function Page(){ return <main/> }
`;
  }

  // Individual page routes
  for (const p of sorted) {
    const route = kebab(p.name);
    files[`next-app/app/${route}/page.tsx`] = `import React from 'react'
import Page from '../../../react/pages/${p.name}'
export default function PageEntry(){ return <Page/> }
`;
  }

  // tsconfig for editor TS in next-app
  files['next-app/tsconfig.json'] = JSON.stringify({ extends: '../tsconfig.json', compilerOptions: { jsx: 'preserve' }, include: ['.'] }, null, 2) + '\n';
  // next.config.js minimal
  files['next-app/next.config.js'] = `/** Auto-generated minimal Next.js config */\nmodule.exports = { reactStrictMode: true }\n`;

  return files;
}
