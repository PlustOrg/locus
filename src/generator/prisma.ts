import { Entity } from '../ast';

export function generatePrismaSchema(db: { entities: Entity[] }): string {
  const header = `generator client {\n  provider = \"prisma-client-js\"\n}\n\ndatasource db {\n  provider = \"postgresql\"\n  url      = env(\"DATABASE_URL\")\n}\n\n`;

  const sorted = [...db.entities].sort((a, b) => a.name.localeCompare(b.name));
  const models = sorted.map(entity => renderModel(entity, sorted)).join('\n\n');
  return header + models + '\n';
}

function renderModel(entity: Entity, _all: Entity[]): string {
  const fields: string[] = [];
  // default id
  fields.push('id Int @id @default(autoincrement())');
  // scalar fields
  for (const f of entity.fields as any[]) {
    let type: string;
    if (f.type.kind === 'list') {
      // Prisma list fields cannot be suffixed with ?; ignore optional flag for list kind
      type = mapType(f.type.of) + '[]';
    } else {
      type = mapType(f.type.name) + (f.type.optional ? '?' : '');
    }
    let line = `${f.name} ${type}`;
    for (const a of f.attributes) {
      if (a.kind === 'unique') line += ' @unique';
      if (a.kind === 'default') line += ' @default(' + renderDefault(a.value) + ')';
      if (a.kind === 'map') line += ` @map(\"${a.to}\")`;
    }
    fields.push(line);
  }
  // relations
  for (const r of entity.relations) {
    if (r.kind === 'has_many') {
      fields.push(`${r.name} ${r.target}[]`);
    } else if (r.kind === 'belongs_to') {
      // Convention: scalar foreign key first, then relation field
      fields.push(`${r.name}Id Int`);
      fields.push(`${r.name} ${r.target} @relation(fields: [${r.name}Id], references: [id])`);
    } else if (r.kind === 'has_one') {
      fields.push(`${r.name} ${r.target}?`);
    }
  }

  return `model ${entity.name} {\n  ${fields.join('\n  ')}\n}`;
}

function mapType(name: string): string {
  switch (name) {
    case 'String': return 'String';
    case 'Text': return 'String';
    case 'Integer': return 'Int';
    case 'Decimal': return 'Decimal';
    case 'Boolean': return 'Boolean';
    case 'DateTime': return 'DateTime';
    case 'Json': return 'Json';
    default: return 'String';
  }
}

function renderDefault(v: any): string {
  if (typeof v === 'string') return `\"${v}\"`;
  if (typeof v === 'number') return String(v);
  if (typeof v === 'boolean') return v ? 'true' : 'false';
  if (v && v.call) {
    const args = (v.args || []).map(renderDefault).join(', ');
    return `${v.call}(${args})`;
  }
  return 'null';
}
