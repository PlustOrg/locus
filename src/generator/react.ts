export function generateReactPage(page: any): string {
  const imports = `import React, { useEffect, useState } from 'react';\n`;
  const compStart = `export default function ${page.name}() {\n`;
  const stateLines = (page.state || []).map((s: any) => `  const [${s.name}, set${capitalize(s.name)}] = useState(${s.default});`).join('\n');
  const onLoad = page.onLoad ? `\n  useEffect(() => {\n    ${page.onLoad}\n  }, []);\n` : '';
  const actions = (page.actions || []).map((a: any) => `  function ${a.name}(${(a.params||[]).join(', ')}) {\n    ${a.body || ''}\n  }`).join('\n\n');
  const uiContent = stripUiWrapper(page.ui) || '<div />';
  const ui = `\n  return (\n    ${uiContent}\n  );\n`;
  const end = `}\n`;
  return [imports, compStart, stateLines, onLoad, actions, ui, end].join('\n');
}

export function generateReactComponent(component: any): string {
  const imports = `import React from 'react';\n`;
  const props = (component.params || []).map((p: any) => `${p.name}: any`).join('; ');
  const propsInterface = props ? `interface ${component.name}Props { ${props} }\n` : '';
  const signature = props ? `(${lowerFirst(component.name)}Props: ${component.name}Props)` : `()`;
  const uiContent = stripUiWrapper(component.ui) || '<div />';
  const comp = `export default function ${component.name}${signature} {\n  return (\n    ${uiContent}\n  );\n}\n`;
  return [imports, propsInterface, comp].join('\n');
}

function capitalize(s: string) { return s.charAt(0).toUpperCase() + s.slice(1); }
function lowerFirst(s: string) { return s.charAt(0).toLowerCase() + s.slice(1); }
function stripUiWrapper(ui?: string): string | undefined {
  if (!ui) return undefined;
  const m = /\bui\s*\{([\s\S]*?)\}$/m.exec(ui.trim());
  if (m) return m[1];
  return ui;
}
