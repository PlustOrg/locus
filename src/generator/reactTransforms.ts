import { UIAttr, UINode } from './uiAstTypes';

export function capitalize(s: string) { return s.charAt(0).toUpperCase() + s.slice(1); }

export function transformUi(ui: string, _state: any[]): string {
  let out = ui;
  out = out.replace(/on:([a-zA-Z]+)/g, (_, ev) => `on${capitalize(ev)}`);
  out = out.replace(/bind:value=\{([A-Za-z_][A-Za-z0-9_]*)\}/g, (_m, v) => {
    const setter = `set${capitalize(v)}`;
    return `value={${v}} onChange={(e) => ${setter}(e.target.value)}`;
  });
  out = out.replace(/<([A-Za-z_][A-Za-z0-9_]*)\s+([^>]*?)for:each=\{\s*([A-Za-z_][A-Za-z0-9_]*)\s+in\s+([^}]+)\}\s*([^>]*)>([\s\S]*?)<\/\1>/g,
    (_m, tag, preAttrs, item, arr, postAttrs, body) => {
      const inner = transformIfElse(body.trim());
      const open = `<${tag} ${preAttrs.trim()} ${postAttrs.trim()}`.replace(/\s+/g, ' ').trim();
      return `{${arr}.map((${item}, index) => (\n  ${open} key={index}>${inner}</${tag}>\n))}`;
    });
  out = out.replace(/<([A-Za-z_][A-Za-z0-9_]*)\s+([^>]*?)for:each=\{\s*([A-Za-z_][A-Za-z0-9_]*)\s+in\s+([^}]+)\}\s*([^>]*)\/>/g,
    (_m, tag, preAttrs, item, arr, postAttrs) => {
      const open = `<${tag} ${preAttrs.trim()} ${postAttrs.trim()}`.replace(/\s+/g, ' ').trim();
      return `{${arr}.map((${item}, index) => (\n  ${open} key={index} />\n))}`;
    });
  out = transformIfElse(out);
  return out;
}

export function transformIfElse(src: string): string {
  const ifMatch = src.match(/<if\s+condition=\{([\s\S]*?)\}\s*>\s*([\s\S]*?)\s*<\/if>/m);
  if (!ifMatch) return src;
  const ifCond = ifMatch[1];
  const ifBody = ifMatch[2];
  const elseifMatch = src.match(/<elseif\s+condition=\{([\s\S]*?)\}\s*>\s*([\s\S]*?)\s*<\/elseif>/m);
  const elseMatch = src.match(/<else\s*>\s*([\s\S]*?)\s*<\/else>/m);
  let middle = '';
  if (elseifMatch) {
    middle = `: ${elseifMatch[1]} ? (\n${elseifMatch[2]}\n) `;
  }
  const end = elseMatch ? `: (\n${elseMatch[1]}\n)` : ': null';
  const ternary = `{${ifCond} ? (\n${ifBody}\n) ${middle}${end}}`;
  return src
    .replace(/<if[\s\S]*?<\/if>/m, '__IF_BLOCK__')
    .replace(/<elseif[\s\S]*?<\/elseif>/m, '')
    .replace(/<else[\s\S]*?<\/else>/m, '')
    .replace('__IF_BLOCK__', ternary);
}

export function renderUiAst(node: any): string {
  if (node.type === 'text') return node.value;
  if (node.type === 'expr') return `{${node.value}}`;
  if (node.type === 'slot') return node.name === 'default' ? '{children}' : `{${node.name}}`;
  if (node.type === 'forEach') {
    const inner = renderElement(node.template.tag, node.template.attrs, node.template.children, true);
    return `{${node.iterable}.map((${node.item}, index) => (\n${inner.replace(/<([A-Za-z0-9_:-]+)([^>]*)\/>/, '<$1$2 key={index} />')}\n))}`;
  }
  if (node.type === 'if') {
    let expr = `{${node.condition} ? (\n${renderChildren(node.consequent)}\n)`;
    if ((node.elif || []).length) for (const e of node.elif) expr += ` : ${e.condition} ? (\n${renderChildren(e.children)}\n)`;
    if (node.else && node.else.length) expr += ` : (\n${renderChildren(node.else)}\n)`; else expr += ' : null';
    expr += '}';
    return expr;
  }
  if (['if','elseif','else'].includes(node.tag)) return renderChildren(node.children||[]);
  return renderElement(node.tag, node.attrs || {}, node.children || [], false);
}

export function renderElement(tag: string, attrs: Record<string, UIAttr>, children: UINode[], _alreadyHasKey: boolean): string {
  const attrStrs: string[] = [];
  for (const [k, v] of Object.entries(attrs)) {
    if (k === 'forEach') continue;
    if (k.startsWith('bind$') && (v as any).kind === 'expr') {
      const prop = k.slice(5);
      const stateVar = (v as any).value;
      const setter = `set${capitalize(stateVar)}`;
      if (prop === 'value') { attrStrs.push(`value={${stateVar}}`, `onChange={(e) => ${setter}(e.target.value)}`); continue; }
      if (prop === 'checked') { attrStrs.push(`checked={${stateVar}}`, `onChange={(e) => ${setter}(e.target.checked)}`); continue; }
      if (prop === 'disabled') { attrStrs.push(`disabled={${stateVar}}`); continue; }
      attrStrs.push(`${prop}={${stateVar}}`, `onChange={(e) => ${setter}(e.target?.value ?? e.target?.checked)}`);
      continue;
    }
    const key = k === 'class' ? 'className' : k;
    if ((v as any).kind === 'string') attrStrs.push(`${key}="${(v as any).value}"`);
    else if ((v as any).kind === 'expr') attrStrs.push(`${key}={${(v as any).value}}`);
  }
  const open = `<${tag}${attrStrs.length ? ' ' + attrStrs.join(' ') : ''}`;
  if (!children || children.length === 0) return `${open} />`;
  const inner = renderChildren(children);
  return `${open}>${inner}</${tag}>`;
}

export function renderChildren(children: UINode[]): string {
  const out: string[] = [];
  for (let i = 0; i < children.length; i++) {
    const cur: any = children[i];
    if (cur && cur.type === 'element' && cur.tag === 'if') {
      const chain = [cur];
      let j = i + 1;
      while (j < children.length) {
        const next: any = children[j];
        if (next && next.type === 'element' && (next.tag === 'elseif' || next.tag === 'else')) { chain.push(next); j++; } else break;
      }
      const ifNode = chain[0];
      const cond = (((ifNode.attrs || {}) as any).condition || { kind: 'expr', value: 'false' }).value;
      const ifBody = (ifNode.children || []).map((c: any) => renderUiAst(c)).join('');
      let expr = `{${cond} ? (\n${ifBody}\n)`;
      for (let k = 1; k < chain.length; k++) {
        const node: any = chain[k];
        if (node.tag === 'elseif') {
          const ec = (((node.attrs || {}) as any).condition || { kind: 'expr', value: 'false' }).value;
          const eb = (node.children || []).map((c: any) => renderUiAst(c)).join('');
          expr += ` : ${ec} ? (\n${eb}\n)`;
        } else if (node.tag === 'else') {
          const eb = (node.children || []).map((c: any) => renderUiAst(c)).join('');
          expr += ` : (\n${eb}\n)`;
        }
      }
      if (!chain.some((n: any) => n.tag === 'else')) expr += ' : null';
      expr += '}';
      out.push(expr);
      i = j - 1;
    } else {
      out.push(renderUiAst(children[i] as any));
    }
  }
  return out.join('');
}
