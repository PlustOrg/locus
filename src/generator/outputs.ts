import { join } from 'path';
import { generatePrismaSchema } from './prisma';
import { generateExpressApi } from './express';
import { generateReactComponent, generateReactPage } from './react';
import { generateCssVariables } from './theme';
import { generateNextApp } from './next';
import { parseToml } from '../config/toml';
import { existsSync, readFileSync } from 'fs';

export interface BuildArtifactsOptions {
  srcDir: string;
  appName?: string; // override computed name
  includeNext?: boolean; // default true
  includeTheme?: boolean; // default true
}

export interface UnifiedLike {
  database: { entities: any[] };
  pages: any[];
  components: any[];
  designSystem?: any;
}

export function withHeader(content: string, kind?: string) {
  const header = '// AUTO-GENERATED by Locus. DO NOT EDIT.' + (kind ? ` (${kind})` : '') + '\n';
  return content.startsWith('// AUTO-GENERATED') ? content : header + content;
}

export function getAppName(srcDir: string): string {
  try {
    const tomlPath = join(srcDir, 'Locus.toml');
    if (existsSync(tomlPath)) {
      const raw = readFileSync(tomlPath, 'utf8');
      const parsed = parseToml(raw);
      const candidate = parsed?._sections?.app?.name;
      if (typeof candidate === 'string') {
        const cleaned = candidate.trim().toLowerCase().replace(/[^a-z0-9-_]/g, '-').replace(/^-+/, '');
        if (cleaned) return cleaned;
      }
    }
  } catch {/* ignore */}
  return 'locus-generated-app';
}

export function buildPackageJson(hasPages: boolean, appName: string): string {
  const pkg = {
    name: appName,
    private: true,
    version: '0.0.0',
    scripts: {
      'dev:api': 'node -r ts-node/register/transpile-only server.ts',
      ...(hasPages ? { 'dev:next': 'next dev next-app' } : {}),
      dev: hasPages ? 'npm run dev:api & npm run dev:next' : 'npm run dev:api',
      'prisma:generate': 'prisma generate',
      build: hasPages ? 'next build next-app' : 'echo "No Next app"',
  start: hasPages ? 'next start next-app' : 'node server.ts',
  'start:api': 'node dist/server.js'
    },
    dependencies: {
      express: '^4.19.2',
      '@prisma/client': '^5.15.0',
      cors: '^2.8.5',
      ...(hasPages ? { react: '^18.3.1', 'react-dom': '^18.3.1', next: '^14.2.0' } : {})
    },
  devDependencies: { prisma: '^5.15.0', 'ts-node': '^10.9.2', typescript: '^5.4.0', '@types/node': '^20.11.0', '@types/react': '^18.2.0' }
  } as any;
  return JSON.stringify(pkg, null, 2);
}

export function buildTsConfig(): string {
  return JSON.stringify({
    compilerOptions: {
      target: 'ES2020',
      module: 'commonjs',
      esModuleInterop: true,
      skipLibCheck: true,
      strict: false,
      types: []
    },
    include: ['**/*.ts', '**/*.tsx']
  }, null, 2);
}

export function buildGeneratedReadme(): string {
  return `# Locus Generated App\n\n## Quick Start\n\nInstall dependencies:\n\n    npm install\n\nGenerate Prisma client (if not yet):\n\n    npm run prisma:generate\n\nStart development (API + Next):\n\n    npm run dev\n\nVisit API: http://localhost:3001 (default)\nVisit App: http://localhost:3000 (if Next pages generated)\n\nBuild production (Next):\n\n    npm run build\n\nRun production server:\n\n    npm start\n\n## Environment Variables\n\nCreate a .env from .env.example and set DATABASE_URL.\nOptional: ENABLE_CORS=1 to enable CORS on the API.\n\n## Notes\n\nThis directory is auto-generated by Locus; manual edits may be overwritten.\n`;
}

export function buildOutputArtifacts(merged: UnifiedLike, opts: BuildArtifactsOptions) {
  if (process.env.LOCUS_TEST_FORCE_GENERATOR_ERROR === '1') {
    throw new Error('Forced generator error (test)');
  }
  const files: Record<string, string> = {};
  // Prisma
  files['prisma/schema.prisma'] = withHeader(generatePrismaSchema(merged.database), 'prisma schema');
  // .env.example (only if not already present later; consumer decides to skip overwrite)
  files['.env.example'] = 'DATABASE_URL=postgresql://user:password@localhost:5432/mydb?schema=public\n';
  // Express (routes + server)
  const routes = generateExpressApi(merged.database.entities as any);
  for (const [rel, content] of Object.entries(routes)) {
    files[rel] = withHeader(content as string, rel.includes('server.ts') ? 'express server' : 'express route');
  }
  // React pages & components
  const pages = [...merged.pages].sort((a, b) => a.name.localeCompare(b.name));
  for (const p of pages) {
    files[`react/pages/${p.name}.tsx`] = withHeader(generateReactPage(p), 'react page');
  }
  const components = [...merged.components].sort((a, b) => a.name.localeCompare(b.name));
  for (const c of components) {
    files[`react/components/${c.name}.tsx`] = withHeader(generateReactComponent(c), 'react component');
  }
  // Theme
  if (opts.includeTheme !== false) {
    const css = withHeader(generateCssVariables(merged.designSystem), 'theme css');
    files['theme.css'] = css;
  }
  // Next
  if (opts.includeNext !== false) {
    const nextFiles = generateNextApp(merged.pages as any);
    for (const [rel, content] of Object.entries(nextFiles)) {
      files[rel] = withHeader(content as string, 'next');
    }
    if (files['theme.css']) {
      files['next-app/public/theme.css'] = files['theme.css'];
    }
  }
  return { files, meta: { hasPages: pages.length > 0 } };
}
