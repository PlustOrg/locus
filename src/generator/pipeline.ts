import crypto from 'crypto';
import { withHeader } from './outputs';

// --- Generation cache & counters ---
let __generationRuns = 0;
const __buildCache = new Map<string, { files: Record<string,string>; meta: any }>();
export function __getGenerationRunCount(){ return __generationRuns; }

// inline react-runtime step extracted to steps/step.reactRuntime.ts
import { integrateExternalSteps } from './steps';


export interface GenerationContext {
  unified: any;
  options: { includeNext?: boolean; includeTheme?: boolean };
  files: Record<string,string>;
  warnings: string[];
  meta: Record<string, any>;
  addFile(path: string, content: string, kind?: string): void;
  addWarning(msg: string): void;
}

export interface GeneratorStep {
  name: string;
  run(ctx: GenerationContext): void;
}


export function createContext(unified: any, options: { includeNext?: boolean; includeTheme?: boolean }): GenerationContext {
  const files: Record<string,string> = {};
  const warnings: string[] = [];
  const meta: Record<string, any> = {};
  return {
    unified,
    options,
    files,
    warnings,
    meta,
    addFile(path, content, kind) {
      if (files[path] != null) {
        warnings.push(`duplicate file skipped: ${path}`);
        return;
      }
      files[path] = withHeader(content, kind);
    },
    addWarning(msg) { warnings.push(msg); }
  };
}

const builtinStepsInline: GeneratorStep[] = []; // replaced by integrateExternalSteps()
export const builtinSteps: GeneratorStep[] = integrateExternalSteps(builtinStepsInline);

export function runPipeline(unified: any, options: { includeNext?: boolean; includeTheme?: boolean } = {}) {
  if (process.env.LOCUS_TEST_FORCE_GENERATOR_ERROR === '1') throw new Error('Forced generator error (test)');
  const cacheEnabled = process.env.LOCUS_BUILD_CACHE === '1';
  let cacheKey: string | null = null;
  if (cacheEnabled) {
    try { cacheKey = crypto.createHash('sha1').update(JSON.stringify(unified, (_k,v)=> typeof v === 'function' ? undefined : v)).digest('hex'); } catch { cacheKey = null; }
    if (cacheKey && __buildCache.has(cacheKey)) {
      const cached = __buildCache.get(cacheKey)!;
      return { files: cached.files, meta: { ...cached.meta, fromCache: true } };
    }
  }
  const ctx = createContext(unified, options);
  __generationRuns++;
  const parallelGen = process.env.LOCUS_PARALLEL_GEN === '1';
  if (!parallelGen) {
    for (const step of builtinSteps) step.run(ctx);
  } else {
    // simple phase grouping for order dependencies
    const phases: string[][] = [
      ['react-runtime','prisma','express','workflows-manifest'],
      ['react-components'],
      ['react-pages'],
      ['theme'],
      ['next'],
      ['warnings-summary'],
      ['manifest']
    ];
    const stepMap = new Map(builtinSteps.map(s=>[s.name,s] as const));
    for (const phase of phases) {
      const runnables = phase.map(n=> stepMap.get(n)!).filter(Boolean);
      awaitAll(runnables.map(s=> () => s.run(ctx))); // run in parallel inside phase
    }
  }
  const result = { files: ctx.files, meta: { ...ctx.meta, warnings: ctx.warnings, fromCache: false } };
  if (cacheKey) __buildCache.set(cacheKey, { files: result.files, meta: result.meta });
  return result;
}

function awaitAll(tasks: Array<() => void | Promise<void>>) {
  return Promise.all(tasks.map(t => Promise.resolve().then(t)));
}
