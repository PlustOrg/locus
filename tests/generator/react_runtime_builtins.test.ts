import { generateReactPage } from '../../src/generator/react';

describe('React generator: built-in runtime component imports', () => {
  it('should import used built-in components, sorted alphabetically', () => {
    const page: any = {
      name: 'HomePage',
      ui: 'ui { <Stack><Button>Click me</Button></Stack> }',
    };
    const knownComponentNames: string[] = [];
    const code = generateReactPage(page, knownComponentNames);
    expect(code).toContain("import { Button, Stack } from '../runtime/_locusRuntime';");
  });

  it('should not import a built-in component if a user component with the same name exists', () => {
    const page: any = {
      name: 'HomePage',
      ui: 'ui { <Stack><Button>Click me</Button><Input /></Stack> }',
    };
    // "Button" is a user-defined component, so it should not be imported from the runtime.
    const knownComponentNames: string[] = ['Button'];
    const code = generateReactPage(page, knownComponentNames);

    // It should import the user's Button component.
    expect(code).toContain("import Button from '../components/Button'");

    // It should import only Input and Stack from the runtime.
    expect(code).toContain("import { Input, Stack } from '../runtime/_locusRuntime';");
    expect(code).not.toContain("Button,");
  });

  it('should not add a runtime import if no built-in components are used', () => {
    const page: any = {
      name: 'HomePage',
      ui: 'ui { <div>Hello</div> }',
    };
    const knownComponentNames: string[] = [];
    const code = generateReactPage(page, knownComponentNames);
    expect(code).not.toContain('../runtime/_locusRuntime');
  });

  it('should handle a mix of user and built-in components and match snapshot for import order', () => {
    const page: any = {
      name: 'ComplexPage',
      ui: 'ui { <Stack><UserCard /><Text>Welcome</Text><Input placeholder="Search..." /></Stack> }',
    };
    const knownComponentNames: string[] = ['UserCard'];
    const code = generateReactPage(page, knownComponentNames);

    const expectedOrder = [
      "import React from 'react';",
      "import { Input, Stack, Text } from '../runtime/_locusRuntime';",
      "import UserCard from '../components/UserCard'",
    ].join('\n');

    expect(code).toContain(expectedOrder);
    expect(code).toMatchSnapshot();
  });
});
