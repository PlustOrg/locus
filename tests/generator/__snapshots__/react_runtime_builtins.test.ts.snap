// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React generator: built-in runtime component imports should handle a mix of user and built-in components and match snapshot for import order 1`] = `
"import React from 'react';
import { Input, Stack, Text } from '../runtime/_locusRuntime';
import UserCard from '../components/UserCard'

export default function ComplexPage() {
  return (
     <Stack><UserCard /><Text>Welcome</Text><Input placeholder="Search..." /></Stack>
  );
}
"
`;
